File: bot/music/queue.py
from __future__ import annotations

import re
from typing import TYPE_CHECKING

import discord
import wavelink

from bot.logger import get_logger
from .player import MusicPlayer
from .track import Track

if TYPE_CHECKING:
    from discord.ext import commands

logger = get_logger(__name__)


async def get_or_create_player(
    guild: discord.Guild,
    voice_channel: discord.VoiceChannel,
    text_channel: discord.abc.Messageable,
    bot: commands.Bot,
) -> MusicPlayer:
    guild_id = str(guild.id)
    if guild_id in bot.queues:
        existing: MusicPlayer = bot.queues[guild_id]
        # Move to new channel if needed
        if existing._wl_player.channel and existing._wl_player.channel.id != voice_channel.id:
            await existing._wl_player.move_to(voice_channel)
        return existing

    wl_player: wavelink.Player = await voice_channel.connect(cls=wavelink.Player, self_deaf=True)
    player = MusicPlayer(guild, wl_player, text_channel, bot)
    bot.queues[guild_id] = player
    return player


async def resolve_tracks(query: str, requested_by: str | None = None) -> list[Track]:
    if re.search(r'open\.spotify\.com/(track|album|playlist)/', query):
        return await _resolve_spotify(query, requested_by)

    return await _resolve_via_wavelink(query, requested_by)


async def _resolve_via_wavelink(query: str, requested_by: str | None) -> list[Track]:
    is_url = query.startswith('http://') or query.startswith('https://')
    search_query = query if is_url else f'ytsearch:{query}'

    try:
        results: wavelink.Search = await wavelink.Playable.search(search_query)
    except Exception as e:
        raise RuntimeError(f'Could not find: {query} — {e}') from e

    if not results:
        raise RuntimeError(f'No results found for: {query}')

    if isinstance(results, wavelink.Playlist):
        return [Track.from_wavelink(t, requested_by) for t in results.tracks[:50]]

    # Single track or search result — return only the first match
    return [Track.from_wavelink(results[0], requested_by)]


_spotify_access_token: str | None = None
_spotify_token_expiry: float = 0.0


async def _ensure_spotify_token() -> bool:
    import os
    import time
    import aiohttp

    global _spotify_access_token, _spotify_token_expiry

    client_id = os.getenv('SPOTIFY_CLIENT_ID')
    client_secret = os.getenv('SPOTIFY_CLIENT_SECRET')
    if not client_id or not client_secret:
        return False

    if time.time() < _spotify_token_expiry:
        return True

    import base64
    credentials = base64.b64encode(f'{client_id}:{client_secret}'.encode()).decode()

    try:
        async with aiohttp.ClientSession() as session:
            async with session.post(
                'https://accounts.spotify.com/api/token',
                headers={
                    'Authorization': f'Basic {credentials}',
                    'Content-Type': 'application/x-www-form-urlencoded',
                },
                data='grant_type=client_credentials',
            ) as resp:
                if not resp.ok:
                    raise RuntimeError(f'Spotify token request failed: {resp.status}')
                data = await resp.json()
                _spotify_access_token = data['access_token']
                _spotify_token_expiry = time.time() + data['expires_in'] - 60
                return True
    except Exception as e:
        logger.error('[Spotify] Failed to get token: %s', e)
        return False


async def _spotify_get(path: str) -> dict:
    import aiohttp

    async with aiohttp.ClientSession() as session:
        async with session.get(
            f'https://api.spotify.com/v1{path}',
            headers={'Authorization': f'Bearer {_spotify_access_token}'},
        ) as resp:
            if not resp.ok:
                raise RuntimeError(f'Spotify API error: {resp.status} {resp.reason} ({path})')
            return await resp.json()


async def _resolve_spotify(url: str, requested_by: str | None) -> list[Track]:
    ok = await _ensure_spotify_token()
    if not ok:
        raise RuntimeError('Spotify support is not configured.')

    track_match = re.search(r'open\.spotify\.com/track/([A-Za-z0-9]+)', url)
    album_match = re.search(r'open\.spotify\.com/album/([A-Za-z0-9]+)', url)
    playlist_match = re.search(r'open\.spotify\.com/playlist/([A-Za-z0-9]+)', url)

    track_names: list[str] = []

    if track_match:
        t = await _spotify_get(f'/tracks/{track_match.group(1)}')
        track_names.append(f'{t["artists"][0]["name"]} {t["name"]}')
    elif album_match:
        data = await _spotify_get(f'/albums/{album_match.group(1)}/tracks?limit=50')
        for t in data['items'][:50]:
            track_names.append(f'{t["artists"][0]["name"]} {t["name"]}')
    elif playlist_match:
        data = await _spotify_get(f'/playlists/{playlist_match.group(1)}/tracks?limit=50')
        for item in data['items'][:50]:
            if item.get('track'):
                t = item['track']
                track_names.append(f'{t["artists"][0]["name"]} {t["name"]}')
    else:
        raise RuntimeError('Unsupported Spotify URL.')

    tracks: list[Track] = []
    for name in track_names:
        try:
            resolved = await _resolve_via_wavelink(name, requested_by)
            tracks.extend(resolved)
        except Exception:
            pass
    return tracks

File: src/music/Queue.js
import { joinVoiceChannel } from '@discordjs/voice';
import { MusicPlayer } from './Player.js';
import { Track } from './Track.js';
import ytDlpExec from 'yt-dlp-exec';

let spotifyAccessToken = null;
let spotifyTokenExpiry = 0;
const spotifyEnabled = !!(process.env.SPOTIFY_CLIENT_ID && process.env.SPOTIFY_CLIENT_SECRET);

async function ensureSpotifyToken() {
  if (!spotifyEnabled) return false;
  if (Date.now() < spotifyTokenExpiry) return true;
  try {
    const credentials = Buffer.from(
      `${process.env.SPOTIFY_CLIENT_ID}:${process.env.SPOTIFY_CLIENT_SECRET}`,
    ).toString('base64');
    const res = await fetch('https://accounts.spotify.com/api/token', {
      method: 'POST',
      headers: {
        Authorization: `Basic ${credentials}`,
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body: 'grant_type=client_credentials',
    });
    if (!res.ok) throw new Error(`Spotify token request failed: ${res.status}`);
    const data = await res.json();
    spotifyAccessToken = data.access_token;
    spotifyTokenExpiry = Date.now() + (data.expires_in - 60) * 1000;
    return true;
  } catch (err) {
    console.warn('[Spotify] Failed to get token:', err.message);
    return false;
  }
}

async function spotifyGet(path) {
  const res = await fetch(`https://api.spotify.com/v1${path}`, {
    headers: { Authorization: `Bearer ${spotifyAccessToken}` },
  });
  if (!res.ok) throw new Error(`Spotify API error: ${res.status} ${res.statusText} (${path})`);
  return res.json();
}

/**
 * Get or create a MusicPlayer for a guild, joining the voice channel.
 * @param {import('discord.js').Guild} guild
 * @param {import('discord.js').VoiceChannel} voiceChannel
 * @param {import('discord.js').TextChannel} textChannel
 * @param {import('discord.js').Client} client
 * @returns {MusicPlayer}
 */
export function getOrCreatePlayer(guild, voiceChannel, textChannel, client) {
  if (client.queues.has(guild.id)) {
    return client.queues.get(guild.id);
  }

  const player = new MusicPlayer(guild.id, client);
  player.textChannel = textChannel;

  const connection = joinVoiceChannel({
    channelId: voiceChannel.id,
    guildId: guild.id,
    adapterCreator: guild.voiceAdapterCreator,
    selfDeaf: true,
  });

  connection.subscribe(player.audioPlayer);
  client.queues.set(guild.id, player);
  return player;
}

/**
 * Resolve a query/URL into Track(s) via yt-dlp or Spotify.
 * @param {string} query
 * @param {string} [requestedBy]
 * @returns {Promise<Track[]>}
 */
export async function resolveTracks(query, requestedBy) {
  // --- Spotify ---
  if (/open\.spotify\.com\/(track|album|playlist)\//.test(query)) {
    return resolveSpotify(query, requestedBy);
  }

  // --- YouTube playlist ---
  if (/youtube\.com\/playlist\?list=/.test(query) || /\?list=/.test(query)) {
    return resolveYouTubePlaylist(query, requestedBy);
  }

  // --- Plain search or direct URL ---
  return resolveYouTube(query, requestedBy);
}

async function resolveYouTube(query, requestedBy) {
  const isUrl = /^https?:\/\//.test(query);
  const target = isUrl ? query : `ytsearch:${query}`;

  try {
    const info = await ytDlpExec(target, {
      dumpSingleJson: true,
      noPlaylist: true,
      noWarnings: true,
      preferFreeFormats: true,
      addHeader: ['referer:youtube.com', 'user-agent:googlebot'],
    });

    if (info.entries) {
      const entry = info.entries[0];
      return [Track.fromYtDlpInfo(entry, requestedBy)];
    }
    return [Track.fromYtDlpInfo(info, requestedBy)];
  } catch (err) {
    throw new Error(`Could not find: ${query} — ${err.message}`);
  }
}

async function resolveYouTubePlaylist(url, requestedBy) {
  try {
    const info = await ytDlpExec(url, {
      dumpSingleJson: true,
      flatPlaylist: true,
      noWarnings: true,
      playlistEnd: 50,
    });

    const entries = info.entries ?? [info];
    return entries.slice(0, 50).map(e => new Track({
      title: e.title ?? 'Unknown',
      url: e.url ?? e.webpage_url ?? `https://www.youtube.com/watch?v=${e.id}`,
      duration: Track.formatDuration(e.duration),
      thumbnail: e.thumbnail ?? e.thumbnails?.[0]?.url ?? null,
      requestedBy,
    }));
  } catch (err) {
    throw new Error(`Failed to load playlist: ${err.message}`);
  }
}

async function resolveSpotify(url, requestedBy) {
  const ok = await ensureSpotifyToken();
  if (!ok) throw new Error('Spotify support is not configured.');

  const trackMatch = url.match(/open\.spotify\.com\/track\/([A-Za-z0-9]+)/);
  const albumMatch = url.match(/open\.spotify\.com\/album\/([A-Za-z0-9]+)/);
  const playlistMatch = url.match(/open\.spotify\.com\/playlist\/([A-Za-z0-9]+)/);

  let trackNames = [];

  if (trackMatch) {
    const t = await spotifyGet(`/tracks/${trackMatch[1]}`);
    trackNames.push(`${t.artists[0].name} ${t.name}`);
  } else if (albumMatch) {
    const data = await spotifyGet(`/albums/${albumMatch[1]}/tracks?limit=50`);
    for (const t of data.items.slice(0, 50)) {
      trackNames.push(`${t.artists[0].name} ${t.name}`);
    }
  } else if (playlistMatch) {
    const data = await spotifyGet(`/playlists/${playlistMatch[1]}/tracks?limit=50`);
    for (const item of data.items.slice(0, 50)) {
      if (item.track) trackNames.push(`${item.track.artists[0].name} ${item.track.name}`);
    }
  } else {
    throw new Error('Unsupported Spotify URL.');
  }

  // Resolve first track immediately; return placeholders for the rest
  const tracks = [];
  for (const name of trackNames) {
    try {
      const [track] = await resolveYouTube(name, requestedBy);
      tracks.push(track);
    } catch {
      // skip unresolvable tracks
    }
  }
  return tracks;
}
